### PR TITLE
[Bootcamp] [AI Accelerators] Convert attention equation to attention call in torch.compile()

### PR DESCRIPTION
**Pull request by kadeng**
Summary:
### Goals

Many attention-based models directly implement scaled dot product attention used by transformer models using component pytorch operators.  Recognize these with FX rewrites and convert them to calls to sdpa(), such that
the optimized implementation is chosen where applicable.

Implementation

Given that the usual scaled dot product attention equation references a shape variable ( because everything gets normalized by a dividing by sqrt(embedding_dim_size), the graph rewrite is not entirely trivial.

It requires a new scale_factor parameter in torch.nn.functional.scaled_dot_product_attention ( not part of this PR, see  discussion https://github.com/pytorch/pytorch/pull/94729#discussion_r1107490896 about earlier version of this PR )

This in turn will allow to rewrite any pattern like

  torch.matmul(query, key.transpose(-2, -1))
        .div(scale_factor)
        .softmax(dim=-1)
        .matmul(value)

into a corresponding (potentially optimized) API call like

  scaled_dot_product_attention(
        query.contiguous(),
        key.contiguous(),
        value.contiguous(),
        attn_mask=None,
        dropout_p=0.0,
        is_causal=False,
        scale_factor=scale_factor,
    )

without knowing the scale factor or query and key shapes at compile time.

### Reviewer Notes:

As discussed with drisspg, this code would be ready for integration into torch._inductor.overrides once the scale_factor parameter is available on torch.nn.functional.scaled_dot_product_attention. Once that is done, the corresponding sections in the unit tests
may be activated again to verify not just the graph rewrite, but also numerical equivalence.

### Followup Tasks:

Once torch.nn.functional.scaled_dot_product_attention has a scale_factor parameter, this rewrite could be activated by configuring it as the sdpa implementation, and the inactive code sections in the unit tests could be made active again.

Test Plan:
Added several new unit tests. These can be executed selectively by running ( Meta Internal Tooling )

  buck2 test @//mode/dev-nosan caffe2/test/inductor:pattern_matcher -- test_sdpa_rewriter

Differential Revision: D43234635

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @soumith @yanboliang @anijain2305 @desertfire @mlazos

### CODE REVIEW

**Comment by kadeng** [comment]
CLA is signed now. I am a Meta employee in Bootcamp.

**Comment by kadeng** [comment]
Update: 
Addressed issues flagged by CI, made torch.nn.functional.scale_factor_dot_product private ( now called torch.nn.functional._scale_factor_dot_product  )

Had to do a force-commit from the patch file after I was unable to merge via export from Meta-internal diff.

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -13819,13 +13819,21 @@
...
+  autogen: _scale_factor_dot_product_attention.out
+
 # This aten function is kept so that we can test the choice function from Python
 - func: _fused_sdp_choice(Tensor query, Tensor key, Tensor value, Tensor? attn_mask=None, float dropout_p=0.0, bool is_causal=False) -> int
   dispatch:
     Meta: _fused_sdp_choice_meta
     CPU, NestedTensorCPU: _fused_sdp_choice_cpp
     CUDA, NestedTensorCUDA: _fused_sdp_choice_cuda
 
+- func: _scale_factor_dot_product_attention_math(Tensor query, Tensor key, Tensor value, Tensor? attn_mask=None, float dropout_p=0.0, bool is_causal=False, Tensor? dropout_mask=None, Scalar scale_factor=1.0) -> (Tensor, Tensor)
```
Comment (drisspg) [inline aten/src/ATen/native/native_functions.yaml:13834]: Adding the optional scale factor Kwarg to the original is something I have planned for the 2.1 release. We don't want to pile in more private functions if we don't have to.

Can you still handle a more limited set of patterns without the generic scale factor?

**Review by drisspg — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (kadeng) [reply aten/src/ATen/native/native_functions.yaml:13834]: > Can you still handle a more limited set of patterns without the generic scale factor?

No, being able to determine the scale factor at execution time is essential as it is applied in every version of scaled_dot_product. 

I can understand that it's kind of ugly to add that private function. I didn't want to mess with an existing public API, otherwise I would not have done that.

But, if you have planned to add the scale_factor kwarg anyway, I can easily modify and simplify this diff so that it uses a configurable function with the same signature as _scale_factor_dot_product_attention. Then the sdp rewrite functionality could be activated later once that function exists. How about that?

Comment (drisspg) [reply aten/src/ATen/native/native_functions.yaml:13834]: Just for my understanding. User code may look like 
``` Python
q * k.T/(q.size(-1) + the rest of sdpa
``` 
You are saying that at FX graph rewrite time though, this will have decomposed into k.transpose().div(<Something>). You can't tell at rewrite time that this scale factor is indeed the one that is baked into SDPA?

Again this will be a null point once the update is landed but trying to see if a smaller unit of work is landable for right after the 2.0 branch cut.

Comment (kadeng) [reply aten/src/ATen/native/native_functions.yaml:13834]: Yes, this has to do with two aspects. 

First of all, at the point of rewrite, I do not have information about the shapes of intermediate tensor results in the fx.Graph. Therefore, the specific value of q.size(-1) is unknown. Symbolically I could, of course, look for .div(sqrt(query.size(-1))), but this would not work reliably for two reasons: First of all, a user could just write .div(sqrt(EMBED_DIM)) where EMBED_DIM is some constant or computed in whatever way imaginable. 

Second, while torch.fx.symbolic_trace converts query.size(-1) to query.size(-1), torch dynamo seems to replace it by query.size(3) (based on shape inference on concrete inputs ). Only it doesn't store that info in the fx graph as far as I could determine ( still a noob on that codebase of course, so maybe I didn't find it, but there is a place for it in (IIRC) node.meta of the fx.Graph, and this was not populated with the neccessary info when I stepped through with a debugger in the corresponding functions in _inductor/overrides.py.

I really need to be able to ensure that q.size(-1) and whatever is divided by are equal. This is easiest at runtime and doesn't really produce any measurable overhead ( e.g. just a few CPU cycles). 

What might be possible is to use the @torch.fx.wrap mechanism to push the decision into python-land. What could happen then are issues popping up from torch dynamo trying to introspect, break up and optimize that function further ( leading to a possible endless recursion of optimization expansions as the replacement pattern would usually also match the replacements fallback implementation ).

Comment (kadeng) [reply aten/src/ATen/native/native_functions.yaml]: I just made the changes I proposed. The PR is way simpler now.

**Review by jansel — CHANGES_REQUESTED** [review]
It doesn't look like this is actually called from anywhere?

This should be done with `pattern_matcher.py`'s patterns which are require `O(1)` passes over the graph rather than `O(num_patterns)`.

**Comment by kadeng** [comment]
> It doesn't look like this is actually called from anywhere?

The original PR called this, but I removed the corresponding code on drisspg's request and was waiting for drisspg's changes to find their way into master. I can now adapt this changeset, so that it will be called from within overrides.py

> This should be done with pattern_matcher.py's patterns which are require O(1) passes over the graph rather than O(num_patterns).

I was not aware of pattern_matcher.py so far. It looks like it is also capable replacing patterns like these, even more efficiently so. But I could not find any examples of comparable pattern complexity in there. Also, there is one advantage of the approach I chose over pattern_matcher.py as it is: The ease of pattern specification as simply parameterized python functions, which does not require an understanding of torch.fx.Graph internals at all (e.g. it is something that could maybe exposed as an API to Pytorch users at some point). I would like to expand the list of replacements further sometime in the future ( if I find time ).  But of course, replacements which are applied *by default* when calling torch.compile should be efficient enough to not slow compilation too much, even in worst case.

I will first adapt the changeset to drisspg's changes and update this PR accordingly with a working version. Then maybe we can have a chat about how to turn these patterns into pattern_matcher.py patterns efficiently.

**Comment by drisspg** [comment]
Another update is that : https://github.com/pytorch/pytorch/pull/95259 has landed which enables the scale kwarg for sdpa

**Comment by jansel** [comment]
> I was not aware of pattern_matcher.py so far. It looks like it is also capable replacing patterns like these, even more efficiently so. But I could not find any examples of comparable pattern complexity in there. Also, there is one advantage of the approach I chose over pattern_matcher.py as it is: The ease of pattern specification as simply parameterized python functions, which does not require an understanding of torch.fx.Graph internals at all (e.g. it is something that could maybe exposed as an API to Pytorch users at some point). I would like to expand the list of replacements further sometime in the future ( if I find time ). But of course, replacements which are applied _by default_ when calling torch.compile should be efficient enough to not slow compilation too much, even in worst case.

All the patterns here seem pretty simple.  Should be easy to convert them over.

**Comment by jansel** [comment]
@kadeng you should be able to generate patterns with something like:
```
def fx_to_pattern(gm):
    """
    Convert an FX graph into a PatternExpr.  This is useful for simple
    patterns that can only match single functions and fixed length lists.
    """
    def not_implemented(*args, **kwargs):
        raise NotImplementedError()

    class Converter(torch.fx.Interpreter):
        call_method = not_implemented
        call_module = not_implemented
        get_attr = not_implemented

        def placeholder(self, target, args, kwargs):
            return KeywordArg(target)

        def call_function(self, target, args, kwargs):
            return CallFunction(target, *args, **kwargs)

    return Converter(gm).run()


def fn_to_pattern(fn, example_inputs):
    from torch.fx.experimental.proxy_tensor import make_fx
    from .decomposition import select_decomp_table
    return fx_to_pattern(make_fx(fn, select_decomp_table())(*example_inputs))
```

The second one will apply all the decomps automatically, so you don't need to try to match a decomposed softmax.

**Comment by kadeng** [comment]
> @kadeng you should be able to generate patterns with something like:

Thanks, I just saw your reply here when I updated my diff. I will try to incorporate that into the most recent version.

**Comment by jansel** [comment]
https://github.com/pytorch/pytorch/pull/97741 includes most of these patterns as a post-grad joint graph pass.
